azure/services/virtualmachineimages/client.go

```diff
@@ -19,45 +19,53 @@ package virtualmachineimages
 import (
 	"context"
 
-	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
-	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v4"
+	"github.com/pkg/errors"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
 // Client is an interface for listing VM images.
 type Client interface {
-	List(ctx context.Context, location, publisher, offer, sku string) (compute.ListVirtualMachineImageResource, error)
+	List(ctx context.Context, location, publisher, offer, sku string) (armcompute.VirtualMachineImagesClientListResponse, error)
 }
 
 // AzureClient contains the Azure go-sdk Client.
 type AzureClient struct {
-	images compute.VirtualMachineImagesClient
+	images armcompute.VirtualMachineImagesClient
 }
 
 var _ Client = (*AzureClient)(nil)
 
-// NewClient creates a new VM images client from auth info.
-func NewClient(auth azure.Authorizer) *AzureClient {
-	return &AzureClient{
-		images: newVirtualMachineImagesClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer()),
+// NewClient creates an azureClient from an Authorizer.
+func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+	c, err := newVirtualMachineImagesClient(auth.SubscriptionID())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new VM extensions client")
 	}
+	return &AzureClient{c}, nil
 }
 
-// newVirtualMachineImagesClient creates a new VM images client from subscription ID, base URI and authorizer.
-func newVirtualMachineImagesClient(subscriptionID, baseURI string, authorizer autorest.Authorizer) compute.VirtualMachineImagesClient {
-	c := compute.NewVirtualMachineImagesClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&c.Client, authorizer)
-	return c
+// newVirtualMachineImagesClient creates a new VM images client from subscription ID.
+func newVirtualMachineImagesClient(subscriptionID string) (armcompute.VirtualMachineImagesClient, error) {
+	credential, err := azidentity.NewDefaultAzureCredential(nil)
```

Comment:
I think this is perhaps the only interesting or controversial change in this PR: is this going to be equivalent auth to the previous approach? I *think* so, but see [DefaultAzureCredential](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#DefaultAzureCredential).

Reply:
How are we now setting the userAgent, correlationID header, and retryAttempts? Basically everything that was part of `SetAutoRestClientDefaults`

What about the regional base URI?

Reply:
AFAICT, the way to set (some of) these options is by passing in an [`arm.ClientOptions`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.5.0/arm#ClientOptions) to each request. This allows us to set retryAttempts and (I think) regional base URI through `cloud.Configuration`.

I haven't figured out how to modify or decorate client request headers yet, and I don't know what to do with with `c.Authorizer = auth` since autorest isn't part of SDKv2 and we're following the recommended default multi-auth. 

Good questions! I'll have clearer answers soon.

Reply:
> I haven't figured out how to modify or decorate client request headers yet

I used a custom [`Transporter`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.5.0/internal/exported#Transporter) to modify the header values. Seems to work.

Reply:
I think I've reimplemented everything from `SetAutoRestClientDefaults`; please take another look when time permits.